operator: Correct changelog to include missing bullet points (#1706)
redpanda-data/redpanda-operator: .changes/operator/v26.2.1.md

## [v26.2.1](https://github.com/redpanda-data/redpanda-operator/releases/tag/operator%2Fv26.2.1) - 2026-07-28
### Added

> * Beta support for StretchCluster that allows deploying operator across multiple kubernetes clusters and creating Redpanda clusters out of RedpandaBrokerPool from different clusters. Requires a valid Redpanda enterprise license.

Reviewer: ```suggestion
* Beta support for stretch clusters. The new StretchCluster resource lets you deploy the operator across multiple Kubernetes clusters and build a single Redpanda cluster from RedpandaBrokerPool resources in different clusters, so your Redpanda cluster can survive the loss of an entire Kubernetes cluster or region. Requires a valid Redpanda Enterprise license.
```

> * Multicluster-aware leader election so multiple operator replicas in the same Kubernetes cluster can take part in the global raft quorum while only the local lease holder votes.

Reviewer: ```suggestion
* Multicluster-aware leader election. You can now run multiple operator replicas in the same Kubernetes cluster for high availability. All replicas take part in the global Raft quorum, but only the replica holding the local lease votes.
```

> * Multicluster PVC unbinder controller that extends stuck-Pod / volume-affinity remediation across all provider clusters.

Reviewer: ```suggestion
* The PVC unbinder controller now works in multicluster mode, so remediation for stuck Pods and volume-affinity problems covers every member cluster, not just the local one.
```

> * Kubeconfig cache on every raft member so a newly elected leader can engage peer clusters without needing live gRPC round-trips to every other operator at promotion time.

Reviewer: ```suggestion
* Every operator replica now caches the kubeconfigs of its peer clusters, so a newly elected leader can start managing peer clusters immediately instead of waiting on live connections to every other operator.
```

> * RedpandaBrokerPool CRD is installed automatically when the operator is deployed in multicluster mode. The RedpandaBrokerPool CRD can manage pools in single K8S cluster.

Reviewer: ```suggestion
* The RedpandaBrokerPool CRD is installed automatically when you deploy the operator in multicluster mode. You can also use it to manage broker pools in a single Kubernetes cluster.
```

> * Per-peer operator Services to support Kubernetes cluster mesh (MCS / Cilium global service) setups for the multicluster raft transport.

Reviewer: ```suggestion
* The operator now creates a Service for each peer, so the multicluster Raft transport works in cluster mesh setups such as Multi-Cluster Services (MCS) or Cilium global services.
```

> * Add Gateway API (HTTPRoute) support to Console CRD, allowing Console to be exposed via Gateway API controllers as an alternative to Ingress. Bumps sigs.k8s.io/gateway-api to v1.5.1.

Reviewer: ```suggestion
* Added Gateway API (HTTPRoute) support to the Console CRD, allowing Console to be exposed via Gateway API controllers as an alternative to Ingress. Bumps sigs.k8s.io/gateway-api to v1.5.1.
```

* Added optional anonymous cluster-shape telemetry: NodePools, stretch clusters, managed Connectors, Redpanda/vectorized cluster and broker counts, broker CPU/memory sizing, Redpanda versions in use, per-cluster TLS/SASL/tiered-storage/console adoption, storage CSI drivers, PVC Unbinder, CRD and supporting-CR counts, and operator feature flags. Licensed clusters additionally report the enterprise license checksum (id_hash) for account correlation. Disabled in builds without an embedded key and via the telemetry.enabled chart value / --disable-telemetry.
* Added a top-level `priorityClassName` value to the operator Helm chart. Set it to a higher-priority PriorityClass so the Redpanda Operator Pod is less likely to be preempted when the cluster is under resource pressure. Defaults to `""` (the cluster's global default PriorityClass, if any).

> * Operator releases now publish the `rpk-k8s` CLI as a managed rpk plugin. On each `operator/v*` release the per-OS/arch `rpk-k8s` binaries are uploaded to the rpk plugins repository (rpk-plugins.redpanda.com) with a generated manifest, so users can install and run it via `rpk k8s` (auto-downloaded, checksum-verified, and cached by rpk on first use). Main subcommand `multicluster bootstrap` can bootstrap TLS certificates and peer configuration for a Redpanda Stretch Cluster. The plugin have useful `rpk-k8s multicluster bundle` subcommand for collecting cross-cluster operator diagnostics (per-peer pod, deployment, TLS, raft status, logs, multi-sample `/metrics`) into a single zip, discovering peers from labeled kubeconfig cache Secrets given any one peer's kubeconfig. A chart-level binding of the operator ServiceAccount to the `metrics-reader` ClusterRole is included so the bundle (and the existing ServiceMonitor) can scrape `/metrics` without 403.

Reviewer: ```suggestion
* Operator releases now publish the `rpk-k8s` CLI as a managed rpk plugin. On each `operator/v*` release the per-OS/arch `rpk-k8s` binaries are uploaded to the rpk plugins repository (rpk-plugins.redpanda.com) with a generated manifest, so users can install and run it with `rpk k8s` (auto-downloaded, checksum-verified, and cached by rpk on first use). Use `rpk k8s multicluster bootstrap` to generate the TLS certificates and peer configuration for a Redpanda stretch cluster. Use `rpk k8s multicluster bundle` to collect cross-cluster operator diagnostics (per-peer Pod, Deployment, TLS, and Raft status, logs, and multiple `/metrics` samples) into a single zip file; it discovers all peers from labeled kubeconfig cache Secrets when you provide any one peer's kubeconfig. The chart also binds the operator ServiceAccount to the `metrics-reader` ClusterRole so the bundle (and the existing ServiceMonitor) can scrape `/metrics` without permission errors.
```

> * Add Broker CRD and minimal controller behind --enable-broker flag

Reviewer: ```suggestion
```

* The `StretchCluster` reconciler can now automatically recover a broker stuck in the "bad_rejoin" crashloop after a region outage by wiping its stale disk: a decommissioned broker whose PVC survived boots from the stale disk, claims a retired identity, is rejected by the cluster, and never becomes Ready. When a broker pod stays not-Ready past a threshold AND its on-disk identity provably collides with the cluster's authoritative broker-uuid view, the operator deletes the pod's PVC and pod so it reschedules onto a fresh disk. This is destructive by design and heavily guarded: it fires only on a confirmed identity collision, sustained unreadiness, a healthy cluster, and zero nodes reported down, wiping at most one disk per reconcile pass. Tune the not-Ready threshold with `--wipe-stale-disk-after` (default 5m); set a negative value to disable the behavior entirely. This stale-disk wipe is distinct from the existing PVCUnbinder controller (`--unbind-pvcs-after`), which frees the PVCs of Pods stuck Pending on lost nodes.

> * Broker custom resource and reconciler (experimental, behind --enable-broker) — per-broker pod/PVC lifecycle with grant-gated serialized rotation, leadership drain, node_id continuity verification, explicit-intent decommission with recommission and last-broker guard, dead-node PV remediation (honors --unbind-pvcs-after when set, 5m default otherwise), and full status conditions

Reviewer: ```suggestion
* Added a `Broker` custom resource and reconciler (experimental, behind the `--enable-broker` flag): per-broker pod/PVC lifecycle with grant-gated serialized rotation, leadership drain, node_id continuity verification, explicit-intent decommission with recommission and last-broker guard, dead-node PV remediation (honors --unbind-pvcs-after when set, 5m default otherwise), and full status conditions.
```

> * Support importing existing Redpanda users into operator management and ongoing credential sync via spec.authentication.syncCredentials for external secret rotation (e.g. ESO).

Reviewer: ```suggestion
* You can now import existing Redpanda users into operator management. Set `spec.authentication.syncCredentials` to keep a user's credentials in sync when an external tool, such as External Secrets Operator (ESO), rotates the password Secret.
```

> * Register the Redpanda Console controller in multicluster operator mode and teach `Console.spec.cluster.clusterRef` to resolve `kind: StretchCluster` so operator-managed Console deployments work in StretchCluster setups. The Console controller previously did not run when the operator was started via `multicluster`, which silently dropped any `cluster.redpanda.com/v1alpha2 Console` CR applied alongside a `StretchCluster`.

Reviewer: ```suggestion
* The Redpanda Console controller is now registered in multicluster operator mode, and `Console.spec.cluster.clusterRef` can resolve `kind: StretchCluster`, so operator-managed Console deployments work in StretchCluster setups. The Console controller previously did not run when the operator was started via `multicluster`, which silently dropped any `cluster.redpanda.com/v1alpha2 Console` CR applied alongside a `StretchCluster`.
```

### Deprecated
* The old broker `DecommissionReconciler` is deprecated. It is no longer reached by `--additional-controllers=decommission` (which now runs the NodePool-aware `StatefulSetDecommissioner`); it remains available only as an opt-in escape hatch via `--additional-controllers=legacy-decommission` (excluded from `all`) for regression fallback and the Cloud Azure path, and will be removed in a future release. The operator logs a warning at startup when it is enabled.
### Removed
* Removed the `nodeWatcher` controller (`RedpandaNodePVCReconciler`). PVC remediation after node loss is now owned solely by the PVCUnbinder (`--unbind-pvcs-after`), which reacts to the actual unschedulable Pod and only touches non-remountable (local/hostPath) volumes. The `nodeWatcher` value for `--additional-controllers` is no longer recognized; deployments should remove it and rely on `--unbind-pvcs-after` instead.
### Fixed

> * Ensure that the field manager "helm" gets removed in migration job to clear out field managers that might be set when migrating from helm to the operator.

Reviewer: ```suggestion
* The Helm-to-operator migration job now removes the `helm` field manager, along with other historic field managers that caused conflicts when the operator updated StatefulSets after a migration from the Helm chart.
```

* Fixed an issue with post-upgrade processing of cluster configuration that resulted in configuration errors due to removed properties across Redpanda versions.

> * Fixed issue with ballast file tuning where the tuning container didn't mount the proper directory needed for propagating the ballast file to the main container.

Reviewer: ```suggestion
* Fixed ballast file tuning: the tuning container now mounts the correct directory, so the ballast file reaches the main Redpanda container.
```

> * Removed more historic field managers that have caused problems with changes to StatefulSets.

Reviewer: ```suggestion
```

* Fixed StretchCluster partial deletion causing cross-cluster resource destruction. When a StretchCluster was deleted from a subset of clusters, the reconciler would destroy resources on surviving clusters and leave the cluster in an unrecoverable state. Added a partial deletion guard that blocks cleanup when the StretchCluster is still active on other clusters, and scoped cross-cluster sync operations (SyncAll, syncBootstrapUser, syncCA) to skip clusters where the StretchCluster is being deleted.
* Fixed a hot reconcile loop on healthy Redpanda clusters where the operator would re-enqueue itself every ~2 seconds instead of settling into the 3-minute periodic requeue interval. The setStatusCondition helper treated a zero rate limit as "force an update" (because time.Since(anything) > 0 is always true), which bumped lastTransitionTime on every pass and triggered a status write for conditions that had not actually changed. Rate-limited heartbeats for License and Configuration conditions are preserved.

> * Fixed an issue where in-use features from an enterprise cluster did not have deterministic sort order and could cause reconciliation storms due to status changes.

Reviewer: ```suggestion
```

> * StretchCluster admin API client now honours the Factory.WithAdminClientTimeout setting on all code paths (previously silently dropped) and the multicluster binary exposes a --cluster-connection-timeout flag to configure it.

Reviewer: ```suggestion
* The StretchCluster admin API client now honors the Factory.WithAdminClientTimeout setting on all code paths (previously silently dropped), and the multicluster binary exposes a --cluster-connection-timeout flag to configure it.
```
